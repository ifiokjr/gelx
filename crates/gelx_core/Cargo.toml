[package]
name = "gelx_core"
version = { workspace = true }
authors = { workspace = true }
categories = ["database"]
edition = { workspace = true }
homepage = { workspace = true }
keywords = ["gel", "database", "typesafe", "checked", "macros"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Utilities for codegen in the `gelx` crate."

[dependencies]
check_keyword = { workspace = true }
gel-derive = { workspace = true }
gel-errors = { workspace = true }
gel-protocol = { workspace = true, features = ["all-types", "with-serde"] }
gel-tokio = { workspace = true, features = ["unstable", "derive"] }
heck = { workspace = true }
indexmap = { workspace = true }
log = { workspace = true }
prettyplease = { workspace = true }
proc-macro2 = { workspace = true }
quote = { workspace = true }
serde = { workspace = true, features = ["derive"] }
syn = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["macros", "rt-multi-thread", "process"] }
typed-builder = { workspace = true }
uuid = { workspace = true, features = ["serde"] }

[dev-dependencies]
assert2 = { workspace = true }
tokio = { workspace = true, features = ["time", "test-util"] }

[features]
with_bigint = []
with_bigdecimal = []
with_chrono = []
with_all = ["with_bigint", "with_bigdecimal", "with_chrono"]
query = []
serde = []
builder = []

[lints]
workspace = true
