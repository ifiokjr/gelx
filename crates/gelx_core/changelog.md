# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.8.1](https://github.com/ifiokjr/gelx/compare/gelx_core-v0.8.0...gelx_core-v0.8.1) - 2025-06-06

### <!-- 1 -->🐛 Bug Fixes

- add annotation to generated `Globals`

### <!-- 2 -->🚜 Refactor

- *(core)* improve codegen code structure

## [0.8.0](https://github.com/ifiokjr/gelx/compare/gelx_core-v0.7.1...gelx_core-v0.8.0) - 2025-06-06

### <!-- 0 -->🎉 Added

- add `Globals` to code generated by the CLI
- support custom scalar types in CLI

### <!-- 1 -->🐛 Bug Fixes

- provide a `crate_module_path` for `typed_builder` macros
- ensure custom scalars are correctly generated for queries and macros

## [0.7.1](https://github.com/ifiokjr/gelx/compare/gelx_core-v0.7.0...gelx_core-v0.7.1) - 2025-06-04

### <!-- 0 -->🎉 Added

- *(gelx)* `builder` should be a default feature

### <!-- 1 -->🐛 Bug Fixes

- *(gelx)* `strum` should not be required as a depedency
- *(gelx)* apply `serde` to all dependencies

### <!-- 6 -->🧪 Testing

- *(gelx_core)* update tests fixtures

## [0.7.0](https://github.com/ifiokjr/gelx/compare/v0.6.0...v0.7.0) - 2025-05-26

### <!-- 0 -->🎉 Added

- *(gelx_core)* generated enums can now be used as `Input` fields
- feat!(gelx_core): remove `create_gel_config` and add `GelxMetadata::gel_config(&self)`
- feat!(gelx_core): remove unused exports `IdentConst` and `Char`
- feat!(gelx_build): add `gelx_build` crate for reading configuration in macros
- add support for postgis via `geo`

### <!-- 1 -->🐛 Bug Fixes

- *(gelx_core)* remove unused `gelx_core` `with_*` features

## [0.6.0](https://github.com/ifiokjr/gelx/compare/v0.5.1...v0.6.0) - 2025-05-24

### <!-- 0 -->🎉 Added

- *(gelx)* update config to `output_path` and generate code in folder rather than file
- *(gelx_core)* add derive macros configuration for structs and enums

### <!-- 2 -->🚜 Refactor

- remove `vfs` and complete refactor
- *(gelx_core)* implement `TryFrom<`&str>` for `GelxMetadata`

### <!-- 6 -->🧪 Testing

- *(gelx)* improve feature testing

## [0.5.0](https://github.com/ifiokjr/gelx/compare/v0.4.0...v0.5.0) - 2025-05-20

### <!-- 0 -->🎉 Added

- _(gelx)_ deprecate `gelx_file` macro
- _(gelx_core)_ add `query_constant_name` metadata option
- _(gelx_core)_ make `exports_alias` customisable
- _(gelx_cli)_ add `--stdout` option for code generation
- _(gelx)_ `gelx!` can now take a custom path to the query

### <!-- 3 -->📚 Documentation

- update installation instructions share readme across crates
- update installation instructions share readme across crates

### <!-- 6 -->🧪 Testing

- _(gelx_cli)_ hardcode `gelx` binary from `devenv` in test

## [0.4.0](https://github.com/ifiokjr/gelx/compare/v0.3.0...v0.4.0) - 2025-05-19

### <!-- 0 -->🎉 Added

- _(cli)_ add `--cwd` to set directory for command
- [**breaking**] code the CLI for generating and checking code
- add `GelxMetadata` and improve everything
- add `gelx_example` crate
- add enum generation from schema
- add `regex` dependency and refactor code generation functions
- add `gelx_cli` crate and integrate `clap` for command-line interface

### <!-- 3 -->📚 Documentation

- small improvements

### <!-- 6 -->🧪 Testing

- fix broken doc tests

## [0.3.0](https://github.com/ifiokjr/gelx/compare/v0.2.1...v0.3.0) - 2025-05-16

### <!-- 0 -->🎉 Added

- [**breaking**] rename `edgedb_codegen` to `gelx`
- [**breaking**] rename `edgedb_codegen_macros` to `gelx_macros`
- [**breaking**] rename `edgedb_codegen_core` to `gelx_core`
- [**breaking**] upgrade all `gel` libraries to latest versions as a replacement for previous `edgedb` dependencies

### <!-- 3 -->📚 Documentation

- improve main readme

## [0.2.0](https://github.com/ifiokjr/gelx/compare/v0.1.2...v0.2.0) - 2024-08-28

### <!-- 0 -->🎉 Added

- [**breaking**] add optional `builder` feature limited to `Input`

### <!-- 3 -->📚 Documentation

- reorder package fields

### <!-- 6 -->🧪 Testing

- improve coverage

## [0.1.2](https://github.com/ifiokjr/gelx/compare/v0.1.1...v0.1.2) - 2024-08-27

### <!-- 0 -->🎉 Added

- _(gelx_core)_ add `prettyplease` format option

### <!-- 1 -->🐛 Bug Fixes

- increase minimum rust version

### <!-- 7 -->⚙️ Miscellaneous Tasks

- update coverage files
- support testing multiple rust versions

## [0.1.1](https://github.com/ifiokjr/gelx/compare/0.1.0...0.1.1) - 2024-08-26

### <!-- 0 -->🎉 Added

- add support for `Range` types

### <!-- 7 -->⚙️ Miscellaneous Tasks

- improve changelog generation

## [0.1.0](https://github.com/ifiokjr/gelx/releases/tag/gelx_core-v0.1.0) - 2024-08-25

### 🎉 Added

- Initial release
