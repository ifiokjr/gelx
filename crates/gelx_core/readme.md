# gelx_core

<br />

> This crate contains the core logic for the `gelx` crate.

## Code generation

Code generation is available for both queries and the database schema.

With the following schema:

```gel
module default {
  type Person {
    required name: str;
  }

  scalar type Genre extending enum<Horror, Comedy, Drama>;

  type Movie {
    required title: str;
    genre: Genre;
    multi actors: Person;
  }
}

module alternative {
	scalar type Awesomeness extending enum<Very, Somewhat, Not>;
}
```

The following generated code is produced:

```rust
use ::gelx::exports as e;

use
#[derive(
	Debug,
	Clone,
	Copy,
	e::serde::Serialize,
	e::serde::Deserialize,
	e::gel_derive::Queryable,
	strum::AsRefStr,
	strum::Display,
	strum::EnumString,
	strum::EnumIs,
	strum::FromRepr,
	strum::IntoStaticStr,
)]
pub enum Genre {
	Horror,
	Comedy,
	Drama,
}

pub mod alternative {
	use super::*;

	#[derive(
		Debug,
		Clone,
		Copy,
		e::serde::Serialize,
		e::serde::Deserialize,
		e::gel_derive::Queryable,
		strum::AsRefStr,
		strum::Display,
		strum::EnumString,
		strum::EnumIs,
		strum::FromRepr,
		strum::IntoStaticStr,
	)]
	pub enum Awesomeness {
		Very,
		Somewhat,
		Not,
	}
}
```
