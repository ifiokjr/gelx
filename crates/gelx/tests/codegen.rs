use std::path::PathBuf;
use std::thread;

use gelx_core::*;
use proc_macro2::Span;
use rstest::fixture;
use rstest::rstest;

fn get_features() -> String {
	let mut features = Vec::new();

	if cfg!(feature = "builder") {
		features.push("builder");
	}

	if cfg!(feature = "query") {
		features.push("query");
	}

	if cfg!(feature = "serde") {
		features.push("serde");
	}

	if cfg!(feature = "strum") {
		features.push("strum");
	}

	if !features.is_empty() {
		features.insert(0, "_");
	}

	features.join("_")
}

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
				let path = format!($($expr,)*);
        settings.set_snapshot_suffix(format!("{path}{}", get_features()));
        let _guard = settings.bind_to_scope();
    }
}

#[fixture]
pub fn testname() -> String {
	thread::current()
		.name()
		.unwrap()
		.split("::")
		.last()
		.unwrap()
		.to_string()
}

#[rstest]
#[case::str("select 'i ❤️ gel'")]
#[case::bool("select true")]
#[case::int64("select 3")]
#[case::float64("select 314e-2")]
#[case::bigint("select 42n")]
#[case::decimal("select 42e+100n")]
#[case::uuid("select <uuid>'a5ea6360-75bd-4c20-b69c-8f317b0d2857'")]
#[case::datetime("select <datetime>'1999-03-31T15:17:00Z'")]
#[case::duration("select <duration>'45.6 seconds'")]
#[case::array("select [1, 2, 3]")]
#[case::tuple("select ('Apple', 7, true)")]
#[case::named_tuple("select (fruit := 'Apple', quantity := 3.14, fresh := true)")]
#[case::set(r#"select {"set", "of", "strings"}"#)]
#[case::empty_set("select <int64>{}")]
#[case::input_shape(
	"select { my_string := RelationshipType.Follow, my_number := 42, several_numbers := {1, 2, \
	 3}, array := [1, 2, 3] };"
)]
#[case::input_shape_with_args(r#"select { hello := "world", custom := <str>$custom }"#)]
#[case::object_shape("select Team {**}")]
#[case::object_shape_with_args(
	"select Team {**} filter .name like <str>$starts_with ++ '%' and .description like '%' ++ \
	 <str>$ends_with;"
)]
#[case::enums("select Account { provider }")]
#[case::types_query(TYPES_QUERY)]
// #[case::range("select range(0, 10)")] // TODO: `Range` doesn't implement `Queryable` yet.
// #[case::bytes("select b'bina\\x01ry'")] // TODO: bytes don't implement `DecodeScalar` yet.
#[tokio::test]
#[rustversion::attr(not(nightly), ignore = "requires nightly")]
async fn codegen_literals(testname: String, #[case] query: &str) -> GelxResult<()> {
	set_snapshot_suffix!("{}", testname);

	let metadata = GelxMetadata::default();
	let relative_path = format!("tests/compile/codegen/{testname}{}.rs", get_features());
	let descriptor = get_descriptor(query, &metadata).await?;
	let code = generate_query_token_stream(&descriptor, "example", query, &metadata, true)?;
	let content = prettify(&code.to_string())?;

	// Check that the snapshot hasn't changed.
	insta::assert_snapshot!(&content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}

#[rstest]
#[case("insert_user")]
#[case("remove_user")]
#[tokio::test]
async fn codegen_files(#[case] path: &str) -> GelxResult<()> {
	set_snapshot_suffix!("{}", path);

	let metadata = GelxMetadata::default();
	let query_path = resolve_path(format!("queries/{path}.edgeql"), Span::call_site())?;
	let relative_path = format!("tests/compile/codegen/{path}{}.rs", get_features());
	let query = tokio::fs::read_to_string(&query_path).await?;
	let descriptor = get_descriptor(&query, &metadata).await?;
	let code = generate_query_token_stream(&descriptor, "example", &query, &metadata, true)?;
	let content = prettify(&code.to_string())?;

	// Check that the snapshot hasn't changed.
	insta::assert_snapshot!(&content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}

const CRATE_DIR: &str = env!("CARGO_MANIFEST_DIR");

async fn prepare_compile_test(content: &str, relative_path: &str) -> GelxResult<()> {
	let is_ci = std::env::var("CI")
		.ok()
		.is_some_and(|v| ["1", "true"].contains(&v.as_str()));

	let path = PathBuf::from(CRATE_DIR).join(relative_path);
	let generated = generate_contents(content)?;

	let should_update = match tokio::fs::read_to_string(&path).await {
		Ok(current) => current != generated,
		Err(_) => true,
	};

	if should_update {
		assert2::assert!(!is_ci, "attempted updating compilation tests in CI");
		tokio::fs::write(&path, generated).await?;
	}

	Ok(())
}

fn generate_contents(content: &str) -> GelxResult<String> {
	let updated = format!("{content}\n\nfn main() {{}}");
	Ok(prettify(&updated)?)
}

#[tokio::test]
async fn can_generate_enums() -> GelxResult<()> {
	set_snapshot_suffix!("enums");

	let relative_path = format!("tests/compile/codegen/enums{}.rs", get_features());
	let tokens = generate_enums(Option::<&str>::None, &GelxMetadata::default(), false).await?;
	let content = prettify(&tokens.to_string())?;
	insta::assert_snapshot!(content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}
#[tokio::test]
async fn can_generate_aliased_enums() -> GelxResult<()> {
	set_snapshot_suffix!("aliased_enums");

	let relative_path = format!("tests/compile/codegen/aliased_enums{}.rs", get_features());
	let tokens = generate_enums(
		Option::<&str>::None,
		&GelxMetadata::builder()
			.features(
				GelxFeatures::builder()
					.query("ssr")
					.builder("ssr")
					.strum("ssr")
					.build(),
			)
			.build(),
		false,
	)
	.await?;

	let content = prettify(&tokens.to_string())?;
	insta::assert_snapshot!(content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}
