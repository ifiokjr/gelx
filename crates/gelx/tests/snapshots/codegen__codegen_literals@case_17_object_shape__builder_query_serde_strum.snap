---
source: crates/gelx/tests/codegen.rs
expression: "&content"
---
pub mod example {
    use ::gelx::exports as e;
    /// Execute the desired query.
    pub async fn query(
        client: &e::gel_tokio::Client,
    ) -> core::result::Result<Vec<Output>, e::gel_errors::Error> {
        client.query(QUERY, &()).await
    }
    /// Compose the query as part of a larger transaction.
    pub async fn transaction(
        conn: &mut e::gel_tokio::Transaction,
    ) -> core::result::Result<Vec<Output>, e::gel_errors::Error> {
        conn.query(QUERY, &()).await
    }
    pub type Input = ();
    #[derive(
        Clone,
        Debug,
        e::serde::Serialize,
        e::serde::Deserialize,
        e::gel_derive::Queryable
    )]
    pub struct OutputWalletsSet {
        pub created_at: e::DateTimeAlias,
        pub id: e::uuid::Uuid,
        pub updated_at: e::DateTimeAlias,
        pub primary: bool,
        pub description: Option<String>,
        pub name: Option<String>,
        pub pubkey: String,
    }
    #[derive(
        Clone,
        Debug,
        e::serde::Serialize,
        e::serde::Deserialize,
        e::gel_derive::Queryable
    )]
    pub struct Output {
        pub slug: String,
        pub id: e::uuid::Uuid,
        pub created_at: e::DateTimeAlias,
        pub updated_at: e::DateTimeAlias,
        pub description: Option<String>,
        pub name: String,
        pub wallets: Vec<OutputWalletsSet>,
    }
    /// The original query string provided to the macro. Can be reused in your codebase.
    pub const QUERY: &str = "select Team {**}";
}
